File: tools/atlas_builder/src/cmake/CMakeParser.kt
/*
 * Copyright(c) 2023-present, Atlas.
 * Distributed under the MIT License (http://opensource.org/licenses/MIT)
 */

package com.spacelion.atlas_builder.cmake

class CMakeParser {
    data class ArgumentFragment(val keywords: String, val startIndex: Int, val hasMultiValue: Boolean)

    fun parseCMakeArgs(statement: String, options: MutableSet<String>, argToValue: MutableMap<String, String>,
                       argToMultiValues: MutableMap<String, List<String>>) {
        for (opt in CMakeTypes.OPTIONS) {
            if (statement.indexOf(opt) != -1) {
                options.add(opt)
            }
        }

        val argumentFragments: MutableList<ArgumentFragment> = mutableListOf()

        for (arg in CMakeTypes.ONE_VALUE_ARGS) {
            val idx = statement.indexOf(arg)
            if (idx != -1) {
                argumentFragments.add(ArgumentFragment(arg, idx, false))
            }
        }

        for (arg in CMakeTypes.MULTI_VALUE_ARGS) {
            val idx = statement.indexOf(arg)
            if (idx != -1) {
                argumentFragments.add(ArgumentFragment(arg, idx, true))
            }
        }

        argumentFragments.sortBy { it.startIndex }

        var lastStartIndex = statement.length
        var fragmentIndex = argumentFragments.size - 1
        while (fragmentIndex >= 0) {
            val fragment = argumentFragments[fragmentIndex]
            --fragmentIndex

            val subStringStart = fragment.startIndex + fragment.keywords.length
            val argLine = normalizeCMakeFunctionArgument(
                statement.substring(subStringStart, lastStartIndex))
            lastStartIndex = fragment.startIndex

            val args = argLine.split(' ')

            if (!fragment.hasMultiValue && args.size == 1) {
                argToValue[fragment.keywords] = args.first()
            } else if (fragment.hasMultiValue && args.isNotEmpty()) {
                argToMultiValues[fragment.keywords] = args
            }
        }
    }

    private fun normalizeCMakeFunctionArgument(argument: String) = argument.trim().removeSuffix("\n")
}

File: tools/AtlasBuilder/CMake/CMakeParser.cs
// Copyright(c) 2023-present, Atlas.
// Distributed under the MIT License (http://opensource.org/licenses/MIT)

namespace AtlasBuilder.CMake;

public static class CMakeParser
{
    public static void ParserCMakeArgs(in string statement, out HashSet<string> options, out Dictionary<string, string> argToValue, out Dictionary<string, List<string>> argToMultiValues)
    {
        options = new HashSet<string>();
        argToValue = new Dictionary<string, string>();
        argToMultiValues = new Dictionary<string, List<string>>();
        List<ArgumentFragment> argumentFragments = new List<ArgumentFragment>();

        foreach (var opt in CMakeTypes.Options)
        {
            int idx = statement.IndexOf(opt, StringComparison.Ordinal);
            if (idx != -1)
            {
                options.Add(opt);
            }
        }
        
        foreach (var arg in CMakeTypes.OneValueArgs)
        {
            int idx = statement.IndexOf(arg, StringComparison.Ordinal);
            if (idx != -1)
            {
                argumentFragments.Add(new ArgumentFragment(arg, idx, false));
            }
        }
        
        foreach (var arg in CMakeTypes.MultiValueArgs)
        {
            int idx = statement.IndexOf(arg, StringComparison.Ordinal);
            if (idx != -1)
            {
                argumentFragments.Add(new ArgumentFragment(arg, idx, true));
            }
        }
        
        argumentFragments.Sort();

        int lastStartIndex = statement.Length;
        int fragmentIndex = argumentFragments.Count - 1;
        while (fragmentIndex >= 0)
        {
            var fragment = argumentFragments[fragmentIndex];
            --fragmentIndex;
            
            var subStringStart = fragment.StartIndex + fragment.Keywords.Length;
            var argLine = statement.Substring(subStringStart, lastStartIndex - 1 - subStringStart);
            lastStartIndex = fragment.StartIndex;
            
            NormalizeCMakeFunctionArgument(ref argLine);
            var args = argLine.Split(' ', StringSplitOptions.RemoveEmptyEntries);
            
            if (!fragment.MultiValue && args.Length == 1)
            {
                argToValue.Add(fragment.Keywords, args[0]);
            }
            else if (fragment.MultiValue && args.Length > 0) 
            {
                argToMultiValues.Add(fragment.Keywords, args.ToList());
            }
        }
    }

    private static void NormalizeCMakeFunctionArgument(ref string argument)
    {
        argument = argument.Trim();
        if (argument.EndsWith('\n'))
        {
            argument = argument.Remove(argument.Length - 1);
        }
    }
    
    private struct ArgumentFragment : IComparable<ArgumentFragment>
    {
        public string Keywords;
        public int StartIndex;
        public bool MultiValue;

        public ArgumentFragment(string keywords, int startIndex, bool multiValue)
        {
            Keywords = keywords;
            StartIndex = startIndex;
            MultiValue = multiValue;
        }
        
        public static bool operator < (ArgumentFragment lhs, ArgumentFragment rhs)
        {
            return lhs.StartIndex < rhs.StartIndex;
        }

        public static bool operator >(ArgumentFragment lhs, ArgumentFragment rhs)
        {
            return lhs.StartIndex > rhs.StartIndex;
        }

        public int CompareTo(ArgumentFragment other)
        {
            return StartIndex.CompareTo(other.StartIndex);
        }
    }
}